typescript/core/src/McapWriter.ts

```diff
@@ -120,23 +126,85 @@ export class McapWriter {
     this.#compressChunk = compressChunk;
   }
 
+  /**
+   * Initializes a new McapWriter for appending to an existing MCAP file. The same `readWrite` will
+   * be used to load indexes out of the existing file, remove the DataEnd and subsequent records,
+   * and then rewrite them when the writer is closed. The existing file must be indexed, since
+   * existing indexes, channel and schema IDs, etc. are reused when appending to the file.
+   *
+   * A writer initialized with this method is already "opened" and does not require a `start()`
+   * call, however it does require an eventual call to `end()` to produce a properly indexed MCAP
+   * file.
+   */
+  static async InitializeForAppending(
+    readWrite: IReadable & ISeekableWriter,
+    options: Omit<McapWriterOptions, "writable" | "startChannelId">,
+  ): Promise<McapWriter> {
+    const reader = await McapIndexedReader.Initialize({ readable: readWrite });
+    await readWrite.seek(reader.dataEndOffset);
+    await readWrite.truncate();
+
+    const writer = new McapWriter({ ...options, writable: readWrite });
+    writer.#appendMode = true;
+    writer.#dataSectionCrc =
+      // Invert the CRC value so we can continue updating it with new data; it will be inverted
+      // again in end()
+      reader.dataSectionCrc != undefined ? crc32Final(reader.dataSectionCrc) : undefined;
+    writer.#chunkIndices = [...reader.chunkIndexes];
+    writer.#attachmentIndices = [...reader.attachmentIndexes];
+    writer.#metadataIndices = [...reader.metadataIndexes];
+
+    if (writer.statistics) {
+      if (reader.statistics) {
+        writer.statistics = reader.statistics;
+      } else {
+        // If statistics calculation was requested, but the input file does not have statistics,
+        // then we can't write them because we don't know the correct initial values
+        writer.statistics = undefined;
+      }
+    }
+
+    writer.#schemas = new Map(reader.schemasById);
+    writer.#writtenSchemaIds = new Set(reader.schemasById.keys());
+    for (const schema of reader.schemasById.values()) {
+      writer.#nextSchemaId = Math.max(writer.#nextSchemaId, schema.id + 1);
+    }
+
+    writer.#channels = new Map(reader.channelsById);
+    writer.#writtenChannelIds = new Set(reader.channelsById.keys());
+    for (const channel of reader.channelsById.values()) {
+      writer.#nextChannelId = Math.max(writer.#nextChannelId, channel.id + 1);
+    }
+
+    return writer;
+  }
+
   async start(header: Header): Promise<void> {
+    if (this.#appendMode) {
+      throw new Error(`Cannot call start() when writer is in append mode`);
+    }
     this.#recordWriter.writeMagic();
     this.#recordWriter.writeHeader(header);
 
-    this.#dataSectionCrc = crc32Update(this.#dataSectionCrc, this.#recordWriter.buffer);
+    if (this.#dataSectionCrc != undefined) {
```

Comment:
Not sure if I fully understand the question – but if I try to comment out this line I get a typescript error on the next line because `#dataSectionCrc` is `number | undefined`. In the normal non-appending mode, it is initialized to `crc32Init()` (`~0`). In append mode, if the dataSectionCrc from the original file is unknown (i.e. zero), we don't attempt to produce an updated value because it would be incorrect (unless we scan back to the beginning of the file and calculate the CRC from scratch).

Reply:
My question was if we can't simply do
```
this.#dataSectionCrc = crc32Init();
```
since this function is only called in the normal non-appending mode.

Reply:
👍 I like that, will remove `crc32Init()` from the member initializer and move it here.